eradiate/scenes/atmosphere/_heterogeneous_new.py

```diff
@@ -0,0 +1,363 @@
+    """
+
+    molecular_atmosphere: Optional[MolecularAtmosphere] = documented(
+        attr.ib(
+            default=None,
+            converter=attr.converters.optional(atmosphere_factory.convert),
+            validator=attr.validators.optional(
+                attr.validators.instance_of(MolecularAtmosphere)
+            ),
+        ),
+        doc="Molecular atmosphere.",
+        type=":class:`MolecularAtmosphere`",
+        default="None",
+    )
+
+    particle_layers: Optional[List[ParticleLayer]] = documented(
+        attr.ib(
+            default=None,
+            converter=attr.converters.optional(
+                lambda value: [ParticleLayer.convert(element) for element in value]
+            ),
+            validator=attr.validators.optional(attr.validators.instance_of(list)),
+        ),
+        doc="Particle layers.",
+        type="List[:class:`ParticleLayer`]",
+        default="None",
+    )
+
+    @molecular_atmosphere.validator
+    @particle_layers.validator
+    def _validate_ids(instance, attribute, value) -> None:
+        ids = []
+        if instance.molecular_atmosphere is not None:
+            ids.append(instance.molecular_atmosphere.id)
+
+        if instance.particle_layers is not None:
+            for particle_layer in instance.particle_layers:
+                if not particle_layer.id in ids:
+                    ids.append(particle_layer.id)
+                else:
+                    raise ValueError(
+                        f"either two particle layers or one particle layer and the molecular atmosphere share the same id ('{ids[-1]}') which is forbidden."
+                    )
+
+    # TODO: the user is probably interested in a shortcut to align the
+    # width of the particle layers with the width of the molecular atmosphere.
+
+    # --------------------------------------------------------------------------
+    #              Spatial extension and thermophysical properties
+    # --------------------------------------------------------------------------
+
+    @property
+    def bottom(self) -> pint.Quantity:
+        bottoms = []
+
+        if self.molecular_atmosphere is not None:
+            bottoms.append(self.molecular_atmosphere.bottom)
+
+        if self.particle_layers is not None:
+            for particle_layer in self.particle_layers:
+                bottoms.append(particle_layer.bottom)
+
+        if len(bottoms) == 0:
+            return 0.0 * ureg.km  # dummy default
+        else:
+            return min(bottoms)
+
+    @property
+    def top(self) -> pint.Quantity:
+        tops = []
+
+        if self.molecular_atmosphere is not None:
+            tops.append(self.molecular_atmosphere.top)
+
+        if self.particle_layers is not None:
+            for particle_layer in self.particle_layers:
+                tops.append(particle_layer.top)
+
+        if len(tops) == 0:
+            return 10.0 * ureg.km  # dummy default
+        else:
+            return max(tops)
+
```

Comment:
A lot of the branching used here would not be necessary if you would have a `components` property returning all components at once. Example:
```python
@property
def components(self):
    result = [self.molecular_atmosphere] if self.molecular_atmosphere else []
    result.extend(self.particle_layers)
    return result
```